cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/async/runnable/RemoveDisksRunnable.java

```diff
@@ -0,0 +1,159 @@
+
+        checkCanRemoveDisks(_brokerIdAndLogdirs, clusterModel);
+
+        OptimizationOptions optimizationOptions = computeOptimizationOptions(clusterModel,
+                false,
+                _kafkaCruiseControl,
+                Collections.emptySet(),
+                _dryRun,
+                _excludeRecentlyDemotedBrokers,
+                _excludeRecentlyRemovedBrokers,
+                _excludedTopics,
+                Collections.emptySet(),
+                false,
+                _fastMode
+        );
+
+        OptimizerResult result = _kafkaCruiseControl.optimizations(clusterModel, _goalsByPriority, _operationProgress, null, optimizationOptions);
+        if (!_dryRun) {
+            _kafkaCruiseControl.executeProposals(
+                    result.goalProposals(),
+                    Collections.emptySet(),
+                    isKafkaAssignerMode(_goals),
+                    0,
+                    0,
+                    1,
+                    1,
+                    SELF_HEALING_EXECUTION_PROGRESS_CHECK_INTERVAL_MS,
+                    SELF_HEALING_REPLICA_MOVEMENT_STRATEGY,
+                    _kafkaCruiseControl.config().getLong(ExecutorConfig.DEFAULT_REPLICATION_THROTTLE_CONFIG),
+                    _isTriggeredByUserRequest,
+                    _uuid,
+                    false
+            );
+        }
+
+        return result;
+    }
+
+    private void checkCanRemoveDisks(Map<Integer, Set<String>> brokerIdAndLogdirs, ClusterModel clusterModel) {
+        for (Map.Entry<Integer, Set<String>> entry : brokerIdAndLogdirs.entrySet()) {
+            Integer brokerId = entry.getKey();
+            Set<String> logDirs = entry.getValue();
+            Broker broker = clusterModel.broker(brokerId);
+            Set<String> brokerLogDirs = broker.disks().stream().map(Disk::logDir).collect(Collectors.toSet());
+            if (!brokerLogDirs.containsAll(logDirs)) {
+                throw new IllegalArgumentException(String.format("Invalid log dirs provided for broker %d.", brokerId));
+            }
+            if (broker.disks().size() < logDirs.size()) {
+                throw new IllegalArgumentException(String.format("Too many log dirs provided for broker %d.", brokerId));
+            } else if (broker.disks().size() == logDirs.size()) {
+                throw new IllegalArgumentException(String.format("No log dir remaining to move replicas to for broker %d.", brokerId));
+            }
+
+            double removedUsage = 0.0;
+            double remainingCapacity = 0.0;
+            double currentUsage = 0.0;
+            for (Disk disk : broker.disks()) {
+                if (logDirs.contains(disk.logDir())) {
+                    removedUsage += disk.utilization();
+                } else {
+                    remainingCapacity += disk.capacity();
+                    currentUsage += disk.utilization();
+                }
+            }
+            double futureUsage = removedUsage + currentUsage;
```

Comment:
Looks like you don't need to record removedUsage and currentUsage separately? The only used var is futureUsage, or total usage.

Reply:
Good catch. Used only futureUsage.